File: csharp/CEnigma/Enigma.cs
namespace CEnigma
{
    public class Enigma
    {
        public Enigma(Rotor reflector, params Rotor[] rotors)
        {
            Reflector = reflector;
            Rotors = rotors;

            Console.WriteLine($"New enigma instance with rotor positions {string.Join(',', Rotors.Select(r => r.Position))}");
        }

        public Rotor Reflector { get; }

        public Rotor[] Rotors { get; }

        public string Encrypt(string input) => new(input.Trim().ToUpper().Select(i => Encrypt(i)).ToArray());

        public char Encrypt(char input)
        {
            var numberOfRotors = Rotors.Length;
            char output = input;

            for(int i = 0; i < numberOfRotors; i++)
            {
                output = Rotors[i].Encode(output);
            }

            output = Reflector.Encode(output);

            for (int i = --numberOfRotors; i >= 0; i--)
            {
                output = Rotors[i].Decode(output);
            }

            AdvanceRotors();

            return output;
        }

        public string Decrypt(string input) => new(input.Trim().ToUpper().Select(i => Decrypt(i)).ToArray());

        public char Decrypt(char input)
        {
            var numberOfRotors = Rotors.Length;
            char output = input;

            for (int i = 0; i < numberOfRotors; i++)
            {
                output = Rotors[i].Encode(output);
            }

            output = Reflector.Decode(output);

            for (int i = --numberOfRotors; i >= 0; i--)
            {
                output = Rotors[i].Decode(output);
            }

            AdvanceRotors();

            return output;
        }

        private void AdvanceRotors()
        {
            Console.Write($"{string.Join(',', Rotors.Select(r => r.Position))} -> ");
            for (int i = 0;i < Rotors.Length; i++)
            {
                if (!Rotors[i].Advance())
                {
                    break;
                }
            }
            Console.WriteLine($"{string.Join(',', Rotors.Select(r => r.Position))}");
        }
    }
}

File: python/Enigma.py
from typing import List
from Rotor import Rotor
from Reflector import Reflector

class Enigma:
    def __init__(self, rotors: List[Rotor], reflector: Reflector) -> None:
        self.rotors = rotors
        self.reflector = reflector

    def encode(self, text: str) -> str:
        output = ''
        for letter in text.upper():
            output += self.encode_letter(letter)
        return output
    
    def decode(self, text: str) -> str:
        output = ''
        for letter in text.upper():
            output += self.decode_letter(letter)
        return output
    
    def encode_letter(self, letter: str) -> str:
        for rotor in self.rotors:
            letter = rotor.encode(letter)
        
        letter = self.reflector.encode(letter)

        for rotor in reversed(self.rotors):
            letter = rotor.decode(letter)

        self.__advance();

        return letter
    
    def decode_letter(self, letter: str) -> str:
        for rotor in self.rotors:
            letter = rotor.encode(letter)
        
        letter = self.reflector.decode(letter)

        for rotor in reversed(self.rotors):
            letter = rotor.decode(letter)

        self.__advance();

        return letter

    def __advance(self) -> None:
        for rotor in self.rotors:
            if not rotor.advance():
                break